addons/stock/models/stock_picking.py

```diff
@@ -370,11 +370,40 @@ class Picking(models.Model):
     immediate_transfer = fields.Boolean(default=False)
     package_level_ids = fields.One2many('stock.package_level', 'picking_id')
     package_level_ids_details = fields.One2many('stock.package_level', 'picking_id')
+    products_availability = fields.Char(compute='_compute_products_availability')
+    products_availability_state = fields.Selection([
+        ('available', 'Available'),
+        ('waiting', 'Should be available on time'),
+        ('late', 'Will not be available at time'),
+    ], compute='_compute_products_availability')
 
     _sql_constraints = [
         ('name_uniq', 'unique(name, company_id)', 'Reference must be unique per company!'),
     ]
 
+    @api.depends('move_lines')
+    def _compute_products_availability(self):
+        inactive_pickings = self.filtered(lambda picking: picking.state in ['cancel', 'draft', 'done'])
+        inactive_pickings.update({
+            'products_availability': '',
+            'products_availability_state': 'available'
+        })
+        active_pickings = (self - inactive_pickings)
+        for picking in active_pickings:
+            worst_moves = picking.move_lines.filtered(lambda move: move.availability_state == 'late')
+            worst_moves = worst_moves or picking.move_lines.filtered(lambda move: move.availability_state == 'partial')
+            worst_moves = worst_moves or picking.move_lines.filtered(lambda move: move.availability_state == 'waiting')
+            if worst_moves:
+                if worst_moves[0].availability_state == 'partial':
+                    picking.products_availability_state = 'late'
+                    picking.products_availability = 'Not Available'
+                else:
+                    picking.products_availability_state = worst_moves[0].availability_state
+                    picking.products_availability = worst_moves[0].availability_indication
+            else:
+                picking.products_availability_state = 'available'
+                picking.products_availability = 'Available'
```

Comment:
All this part is duplicated in mrp. Clean up to have a common method on stock_move